### ISSUE TITLE
Emacs was downgraded, making it necessary to recompile Transient

### ISSUE THREAD

**Issue by namasme** [magit/transient#308]
I run Doom Emacs on version 29.1, and after a recent upgrade I started getting errors from Magit/Transient, allegedly due to versions mismatch. I followed the instructions to uninstall and reinstall Magit, but the issue is still there. Specifically:

1. Removed `magit` from `init.el`.
2. Ran `doom sync` and watched it uninstall `magit`.
3. Restarted emacs and verified `magit` commands were not available.
4. Reintroduced `magit` in `init.el` and ran `doom sync`.
5. Restarted emacs to find `magit` is installed again.

This is the output from `toggle-debug-on-error`:
```
Debugger entered--Lisp error: (cl-assertion-failed ((>= emacs-major-version 30) "Emacs was downgraded, making it necessary to recompile Transient"))
  apply(debug (error (cl-assertion-failed ((>= emacs-major-version 30) "Emacs was downgraded, making it necessary to recompile Transient"))))
  transient--exit-and-debug(error (cl-assertion-failed ((>= emacs-major-version 30) "Emacs was downgraded, making it necessary to recompile Transient")))
  cl--assertion-failed((>= emacs-major-version 30) "Emacs was downgraded, making it necessary to recompile Transient" nil nil)
  transient--wrap-command()
  transient--pre-command()
```

And `magit-version`:
```
Magit 4881835 [>= 3.3.0.50-git], Transient 0.7.5, Git 2.45.1, Emacs 29.1, darwin
```

I also checked that `emacs-major-version` reports 29, which should be < 30. From what I understood (I'm not an Elisp expert) `static-if` is evaluated at compile time, but I'm not sure why the Emacs version would be different at that time.

Any hints as to what I may be doing wrong here?

**Comment by tarsius** [comment]
> From what I understood (I'm not an Elisp expert) `static-if` is
> evaluated at compile time, but I'm not sure why the Emacs version
> would be different at that time.

Yes, that's how that macro works.  One branch is chosen at compile-time and the other is discarded.  The latter helps avoid warnings about things that only exist in certain releases.

Even though you are surprised by it, in all likelihood Emacs >= 30 was used when compiling. 

> Any hints as to what I may be doing wrong here?

If you don't want to use Emacs 30, find the instance of that and remove it.

You could also `M-x byte-compile-file` this library while using Emacs 29.

**Comment by namasme** [comment]
Thanks for your reply!

> Even though you are surprised by it, in all likelihood Emacs >= 30 was used when compiling.
> If you don't want to use Emacs 30, find the instance of that and remove it.

I'm certainly surprised by it because I can't find any traces of an Emacs 30 installed in my box:

```
$ which -a emacs # just one match
/opt/homebrew/bin/emacs

$ emacs --version
GNU Emacs 29.1
Copyright (C) 2023 Free Software Foundation, Inc.
GNU Emacs comes with ABSOLUTELY NO WARRANTY.
You may redistribute copies of GNU Emacs
under the terms of the GNU General Public License.
For more information about these matters, see the file named COPYING.

$ doom info # Doom only knows about 29.1
generated  Sep 18, 2024 12:37:53
system     MacOS 14.6 Darwin 23.6.0 arm64
emacs      29.1 ~/.emacs.d/
doom       3.0.0-pre PROFILE=_@0 HEAD -> master, origin/master, origin/HEAD
           7bdf7cf7c 2023-09-12 21:30:56 +0200 ~/.doom.d/
shell      /bin/zsh
features   ACL GMP GNUTLS JSON LIBXML2 MODULES NATIVE_COMP NOTIFY KQUEUE PDUMPER
           SQLITE3 THREADS TOOLKIT_SCROLL_BARS TREE_SITTER XIM ZLIB
traits     batch server-running envvar-file custom-file
modules    [...]
packages   [...]
```

> You could also M-x byte-compile-file this library while using Emacs 29.

I also tried this by running it on `transient/lisp/transient.el` but got the following error:

> Compiling file /Users/nacho.mas/.emacs.d/.local/straight/repos/transient/lisp/transient.el at Wed Sep 18 12:32:37 2024
In end of data: transient.el:2458:4: Warning: the function ‘static-if’ is not known to be defined.

Is that expected?

**Comment by tarsius** [comment]
>     In end of data: transient.el:2458:4: Warning: the function ‘static-if’ is not known to be defined.
>
> Is that expected?

Sounds like the `compat` package might be outdated. It should be at `30.0.0.0`.

**Comment by namasme** [comment]
According to `M-x list-packages` that's the version I have:

```
compat                         30.0.0.0       available    gnu      Emacs Lisp Compatibility Library
```

**Comment by tarsius** [comment]
That says "available", not "installed".

**Comment by namasme** [comment]
You're right — sorry about that. 

For future reference, I found this Doom-specific issue https://github.com/doomemacs/doomemacs/issues/7939#issuecomment-2223570074 and followed the advice to pin the `compat` package to v30.0.0.0's commit hash, but that still didn't solve the issue.

Then found this discussion https://github.com/minad/corfu/discussions/482#discussioncomment-10062153 and pulled as suggested, followed by `doom sync -u`, and now everything's working again.

I'm not entirely sure which of those steps was the solution, or whether the sequencing was important, but hopefully this should narrow it down for anyone facing the same problem.

Closing for now since the problem is solved for me, and thanks again for the support 🙂
